### ISSUE TITLE
Plots profile or cross section of survey points

### ISSUE THREAD

**Issue by marcoalicera** [PANOimagen/profiletool#14]
I can see that the plugin is prepared to produce a plot from points also, but I miss some functionalities that could allowed a similar workflow to the plugin ProfileFromPoints, which will be nice.

When I load a vector layer with points the plugin crashed.

I tried with 2.5D points, it seems is not supproted:

              File "C:/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\profiletool\tools\ptmaptool.py", line 127, in leftClicked
              self.profiletool.updateProfilFromFeatures(result[0], result[1])
              File "C:/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\profiletool\tools\profiletool_core.py", line 148, in updateProfilFromFeatures
              self.updateProfil(pointstoDraw, False, plotProfil)
              File "C:/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\profiletool\tools\profiletool_core.py", line 174, in updateProfil
              self.profiles[i], _, _ = DataReaderTool().dataVectorReaderTool(self.iface, self.toolrenderer.tool, self.profiles[i], self.pointstoDraw, float(self.dockwidget.mdl.item(i,4).data(QtCore.Qt.EditRole)) )
              File "C:/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\profiletool\tools\dataReaderTool.py", line 259, in dataVectorReaderTool
              projectedpoints = self.interpolateNodeofPolyline(geominlayercrs,projectedpoints)
              File "C:/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\profiletool\tools\dataReaderTool.py", line 347, in interpolateNodeofPolyline
              lenpoly = geom.lineLocatePoint(qgis.core.QgsGeometry.fromPointXY(vertexpoint))
             TypeError: QgsGeometry.fromPointXY(): argument 1 has unexpected type 'QgsPoint'

and with 2D points, but I am not sure how to select the elevation data field and a set of points:

              File "C:/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\profiletool\tools\ptmaptool.py", line 127, in leftClicked
              self.profiletool.updateProfilFromFeatures(result[0], result[1])
              File "C:/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\profiletool\tools\profiletool_core.py", line 148, in updateProfilFromFeatures
              self.updateProfil(pointstoDraw, False, plotProfil)
              File "C:/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\profiletool\tools\profiletool_core.py", line 174, in updateProfil
              self.profiles[i], _, _ = DataReaderTool().dataVectorReaderTool(self.iface, self.toolrenderer.tool, self.profiles[i], self.pointstoDraw, float(self.dockwidget.mdl.item(i,4).data(QtCore.Qt.EditRole)) )
              File "C:/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\profiletool\tools\dataReaderTool.py", line 259, in dataVectorReaderTool
              projectedpoints = self.interpolateNodeofPolyline(geominlayercrs,projectedpoints)
              File "C:/User/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\profiletool\tools\dataReaderTool.py", line 332, in interpolateNodeofPolyline
              projectedpoints = np.append(projectedpoints,[[0,polyline[0].x(), polyline[0].y(), -1,0,projectedpoints[0][5],polyline[0].x(), polyline[0].y(),projectedpoints[0][8] ]], axis = 0)
             IndexError: list index out of range`

One important feature I miss is being able to select the points in the vector layer. So, I have a selected polyline and a set of selected points to do the profile, otherwise I should slpit the point layer for each polyline I want to produce the profile.

**Comment by AnaPC** [comment]
Javier,

Thank you for your work on this plugin. In my opinion, it is so useful that it should by now be integrated in QGIS.

I have a question related to the one from @marcoalicera. In my experience, until about version 4.0.2 the plugin worked with vector point data very well. You would choose the vector layer, and then the "field" to use for plotting, and it would appear in the plot just as DTM profiles do. This was very convenient to compare survey point data with pre-existing DTMs. 

In the newer versions, the plugin crashes every time I try to use it over vector data. I have tried in different computers, in both QGIS 2.18 and 3.x versions, but keep having the same problem. Could you confirm that this is the general behavior now? If that is the case, are there plans to fix the issue? Or am I doing something wrong?

At this point I am working on QGIS 2.18 with an old profile tool plugin version. But it would be great to be able to move on to QGIS 3.4. sometime soon!

Regards,

**Comment by AnaPC** [comment]
Thank you!

**Comment by AnaPC** [comment]
Javier,

Thank you very much for dealing with the issue so fast. I have updated my plugin version to the latest available (4.1.4), and tested the fix. I am using QGIS 2.18.28. Here is my experience:

A) First I tried the profile tool over a raster DTM.  It worked perfectly.

B) Then I tried it over a point layer. The tool created the profiles alright, so it seems that you have managed to get the feature back. However, there are two weird things happening when using a point layer:

1. Every time I click on the screen to select the profile starting point, I get a Python error. It is copied below in case it helps. This does not happen when only rasters are loaded to the profile tool. It stops appearing if I remove all point layers from the profile window. 

2. Once a point layer has been added to the profile tool, the profile window cannot be closed anymore. The "X" button does not respond. This does not get fixed by removing the point layer from the profile list. Also, it does not happen when only raster layers have been used. I have tried a number of times and some times QGIS has crashed, but it has not been consistent. Mostly, it is just the profile window refusing to close.

At this point I don't have QGIS 3 installed in my computer anymore, but can install it later today and try the plugin there.

Regards,

--------------
An error has occurred while executing Python code: 

AttributeError: 'NoneType' object has no attribute 'asPoint' 
Traceback (most recent call last):
  File "C:/Users/anap/.qgis2/python/plugins\profiletool\tools\ptmaptool.py", line 116, in leftClicked
    self.profiletool.updateProfil(self.pointstoDraw)
  File "C:/Users/anap/.qgis2/python/plugins\profiletool\tools\profiletool_core.py", line 174, in updateProfil
    self.profiles[i], _, _ = DataReaderTool().dataVectorReaderTool(self.iface, self.toolrenderer.tool, self.profiles[i], self.pointstoDraw, float(self.dockwidget.mdl.item(i,4).data(QtCore.Qt.EditRole)) )
  File "C:/Users/anap/.qgis2/python/plugins\profiletool\tools\dataReaderTool.py", line 248, in dataVectorReaderTool
    projectedpoints.append([distline,pointprojected.asPoint().x(), pointprojected.asPoint().y(),distpoint,0 ,interptemp, featPnt.geometry().asPoint().x(),featPnt.geometry().asPoint().y(),featPnt ])
AttributeError: 'NoneType' object has no attribute 'asPoint'


Python version: 2.7.14 (v2.7.14:84471935ed, Sep 16 2017, 20:25:58) [MSC v.1500 64 bit (AMD64)] 
QGIS version: 2.18.28 Las Palmas, aa30730ac6 

Python Path:
C:/OSGEO4~1/apps/qgis-ltr/./python/plugins\processing
C:/Users/anap/.qgis2/python/plugins\ntv2_transformations
C:\Users\anap\.qgis2\python\plugins\GeoCoding\libs
C:/OSGEO4~1/apps/qgis-ltr/./python
C:/Users/anap/.qgis2/python
C:/Users/anap/.qgis2/python/plugins
C:/OSGEO4~1/apps/qgis-ltr/./python/plugins
C:\OSGEO4~1\apps\orfeotoolbox\python
Z:\Projects\SEL Monitoring\UAV\2018 Annual Report Figures
C:\OSGEO4~1\bin\python27.zip
C:\OSGEO4~1\apps\Python27\DLLs
C:\OSGEO4~1\apps\Python27\lib
C:\OSGEO4~1\apps\Python27\lib\plat-win
C:\OSGEO4~1\apps\Python27\lib\lib-tk
C:\OSGEO4~1\bin
C:\OSGEO4~1\apps\Python27
C:\OSGEO4~1\apps\Python27\lib\site-packages
C:\OSGEO4~1\apps\Python27\lib\site-packages\jinja2-2.7.2-py2.7.egg
C:\OSGEO4~1\apps\Python27\lib\site-packages\markupsafe-0.23-py2.7-win-amd64.egg
C:\OSGEO4~1\apps\Python27\lib\site-packages\win32
C:\OSGEO4~1\apps\Python27\lib\site-packages\win32\lib
C:\OSGEO4~1\apps\Python27\lib\site-packages\Pythonwin
C:\OSGEO4~1\apps\Python27\lib\site-packages\Shapely-1.2.18-py2.7-win-amd64.egg
C:\OSGEO4~1\apps\Python27\lib\site-packages\wx-2.8-msw-unicode
C:\OSGEO4~1\apps\Python27\lib\site-packages\xlrd-0.9.2-py2.7.egg
C:\OSGEO4~1\apps\Python27\lib\site-packages\xlwt-0.7.5-py2.7.egg
C:/Users/anap/.qgis2//python
C:\Users\anap\.qgis2\python\plugins\DigitizingTools\tools

**Comment by RobertoPanoi** [comment]
We have not updated the plugin in the QGIS repository yet.
You can install the new version manually, if you wish while updating, so you can use it.
__QGIS 3__: Download the zip file ([profiletool.zip](https://github.com/PANOimagen/profiletool/files/2836010/profiletool.zip)) and from the plugins panel you can upload the zip directly to install it.
When it is updated you can download it directly as usual.
Anyway, we recommend upgrading to QGIS 3, since it is already LTR version (QGIS 3.4.4)

Best regards

**Comment by AnaPC** [comment]
Thank you for the clarification. And I apologise for the confusion!

Tomorrow I will reinstall QGIS 3.4 and let you know how it goes.

Regards,

On Wednesday, February 6, 2019, RobertoCSH <notifications@github.com> wrote:
> We have not updated the plugin in the QGIS repository yet.
> You can install the new version manually, if you wish while updating, so
you can use it.
> QGIS 3: Download the zip file (profiletool.zip) and from the plugins
panel you can upload the zip directly to install it.
> When it is updated you can download it directly as usual.
> Anyway, we recommend upgrading to QGIS 3, since it is already LTR version
(QGIS 3.4.4)
>
> Best regards
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.<
https://ci5.googleusercontent.com/proxy/ycRtjsDORfRGjFgCKGP7o6ZxZu48ajlZqKnXPbUViryo1gV4pYfa7_R5areYKygvfswNad41HNYVUyMVsjf5wu2glD4kuGlqTAlcWo9VQdsTrQsiWfuQPRXGl5WiJ7dYfJARZ_34GkGLzMQF232nDjKpaw=s0-d-e1-ft#https://github.com/notifications/beacon/AtEhJMrRMq9ZQjrT3EwTr61yOR6Zxj4-ks5vKraRgaJpZM4Z3NSh.gif>

**Comment by AnaPC** [comment]
Good morning,

I downloaded the zip file, installed it as you directed, and tested it in Qgis 3.4.4. It works great! I tried it over DTMs, over point vector layers, and also using both things at the same time. The tool worked as expected and there were no error/warning messages.

I use the tool pretty frequently, so I will reach out if any issue appears.

Thank you for the quick fix! This is a great job.

Regards,
